library/core/src/mem/mod.rs

```diff
@@ -1129,6 +1129,33 @@ impl<T> fmt::Debug for Discriminant<T> {
 /// assert_eq!(mem::discriminant(&Foo::B(1)), mem::discriminant(&Foo::B(2)));
 /// assert_ne!(mem::discriminant(&Foo::B(3)), mem::discriminant(&Foo::C(3)));
 /// ```
+///
+/// Note that it is *undefined behavior* to [`transmute`] from [`Discriminant`] to a primitive.
+/// To access the value of a discriminant with a primitive representation, use pointer casts instead:
+/// ```
+/// #[repr(u8)]
+/// enum Enum {
+///     Unit,
+///     Tuple(bool),
+///     Struct { a: bool },
+/// }
+///
+/// impl Enum {
+///     fn discriminant(&self) -> u8 {
+///         unsafe { *(self as *const Self as *const u8) }
```

Comment:
We're trying to avoid `as` on pointers now, where possible, in favour of methods.  So please change this to something using `cast` & friends, maybe
```suggestion
///         unsafe { <*const _>::from(self).cast::<u8>().read() }
```

Or you could consider something like
```rust
    pub fn discriminant(&self) -> &u8 {
        unsafe { NonNull::from(self).cast::<u8>().as_ref() }
    }
```

Reply:
Oh, and it would be good to have a comment on the `unsafe` block repeating why it's sound.  Maybe

```
// SAFETY: because `Self` is marked `repr(u8)`,
// its layout is a `repr(C)` `union` between `repr(C)` structs,
// each of which has the `u8` discriminant as its first field,
// so we can read the discriminant without offsetting the pointer.
```

Reply:
TIL that `<*const _>::from(reference)` works. Since there is no `impl From<&T> for *const T`, it probably selects `From<*const T> for *const T` and then coerces the parameter?

We should probably document the "canonical" way to convert a reference to a raw pointer when coercion isn't an option somewhere. So far I've seen `reference as *const _`, `ptr::addr_of!(*reference)` and now `<*const _>::from(reference)`.

Reply:
I've removed the `as` cast on pointers now, but kept `*ptr` over `ptr.read()` since `u8` is `Copy` and that seems "safer" to me.